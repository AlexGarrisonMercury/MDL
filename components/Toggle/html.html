<label class="toggle toggle--large" role="switch" aria-checked="false">
  <input class="toggle__input" type="checkbox" />
  <span class="toggle__track" aria-hidden="true">
    <span class="toggle__knob">X</span>
    <span class="toggle__label-on">On</span>
    <span class="toggle__label-off">Off</span>
  </span>
</label>
