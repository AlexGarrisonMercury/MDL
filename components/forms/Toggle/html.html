<div class="toggle">
  <label class="toggle__switch">
    <input type="checkbox">
  </label>
</div>
