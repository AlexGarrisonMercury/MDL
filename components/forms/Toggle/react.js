import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames/bind'
import s from './style.module.scss'

const cx = classNames.bind(s)

const Toggle = ({
  checked, disabled, icon, onChange, large,
}) => (
  <div
    aria-checked={checked}
    className={cx(['toggle', { 'toggle--large': large }])}
    onClick={() => !disabled && onChange()}
    onKeyPress={() => !disabled && onChange()}
    role="switch"
    tabIndex={0}
  >
    <input
      checked={checked}
      className={s.toggle__input}
      disabled={disabled}
      readOnly
      tabIndex="-1"
      type="checkbox"
    />
    <span className={s.toggle__track} aria-hidden="true">
      <span className={s.toggle__knob} dangerouslySetInnerHTML={{ __html: icon }} />
      {large && (
        <>
          <span className={s['toggle__label-on']}>On</span>
          <span className={s['toggle__label-off']}>Off</span>
        </>
      )}
    </span>
  </div>
)

Toggle.propTypes = {
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  onChange: PropTypes.func,
  large: PropTypes.bool,
}

Toggle.defaultProps = {
  checked: false,
  disabled: false,
  icon: null,
  onChange: null,
  large: false,
}

export default Toggle
