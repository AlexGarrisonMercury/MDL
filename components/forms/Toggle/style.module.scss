@import "styles/colors.scss";


.toggle{

  $size: 14px;
  align-items: center;
  align-content: center;
  position: relative;
  height: $size;
  width: $size;
  flex-shrink: 0;

  &__switch {
    position: relative;
    display: inline-block;
  }
  
  &__switch input { 
    opacity: 0;
    width: 0;
    height: 0;
  }
  
  &__slider {
    position: absolute;
    cursor: pointer;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: $color-gray;
    -webkit-transition: .4s;
    transition: .4s;
    height: 8px;	
    width:24px;
    border-radius:34px;
  }
  
  &__slider:before {
    position: absolute;
    content: "";
    height: 16px;
    width: 16px;
    left: -7px;
    top: -5px;
    bottom: 4px;
    background-color: $color-onyx;
    -webkit-transition: .4s;
    transition: .4s;
  }

  &__input:checked + &__slider:before {
    background-color: $color-limerick;
    left: -10px;
  }
  
  &__input:checked + &__slider {
    background: $color-gray;
  }
  
  &__input:focus + &__slider {
    box-shadow: 0 0 1px $color-onyx;
  }
  
  &__input:checked + &__slider:before {
    -webkit-transform: translateX(26px);
    -ms-transform: translateX(26px);
    transform: translateX(26px);
  }
  
  &__slider:before {
    border-radius: 50%;
  }

}