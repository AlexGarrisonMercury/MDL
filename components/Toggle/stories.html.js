import { storiesOf } from '@storybook/html'
import html from './html.html'

storiesOf('Toggle', module).add('default', () => html)
