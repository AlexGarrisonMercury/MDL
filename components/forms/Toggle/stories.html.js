import { storiesOf } from '@storybook/html'
import html from './html.html'

storiesOf('Forms/Toggle', module).add('default', () => html)
