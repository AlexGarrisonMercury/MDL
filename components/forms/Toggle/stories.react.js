import React from 'react'
import { storiesOf } from '@storybook/react'
import { withKnobs, boolean, select } from '@storybook/addon-knobs'
import { withState } from '@dump247/storybook-state'
import Toggle from './react'

storiesOf('Toggle', module)
  .addDecorator(withKnobs)
  .add(
    'fully controlled',
    withState({ checked: true })(({ store }) => {
      const disabled = boolean('Disabled', false)
      return (
        <Toggle
          checked={store.state.checked}
          disabled={disabled}
          onChange={() => store.set({ checked: !store.state.checked })}
        />
      )
    }),
  )
  .add(
    'large',
    withState({ checked: true })(({ store }) => {
      const disabled = boolean('Disabled', false)
      return (
        <Toggle
          checked={store.state.checked}
          disabled={disabled}
          onChange={() => store.set({ checked: !store.state.checked })}
          large
        />
      )
    }),
  )
