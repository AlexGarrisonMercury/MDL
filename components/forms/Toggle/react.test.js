import React from 'react'
import { mount } from 'enzyme'
import { JSDOM } from 'jsdom'
import Toggle from './react'

const jsdom = new JSDOM('<!doctype html><html><body><div id="app">Mount here</div></body></html>')
const { window } = jsdom

describe('Toggle', () => {
  let props
  let wrapper

  beforeEach(() => {
    props = {
      checked: false,
      onChange: jest.fn(),
      name: 'toggle',
    }

    wrapper = mount(<Toggle {...props} />, {
      attachTo: window.document.querySelector('#app'),
    })
  })

  describe('with snapshots', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('with user actions', () => {
    it('should display the when clicking the toggle button', () => {
      wrapper.find('Toggle').simulate('click')
      expect(props.onChange).toHaveBeenCalledTimes(1)
    })
  })
})
