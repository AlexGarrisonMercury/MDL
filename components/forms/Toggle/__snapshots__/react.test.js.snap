// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle with snapshots should render correctly 1`] = `
<Toggle
  checked={true}
  name="toggle"
  onChange={[MockFunction]}
  onClick={null}
  size={null}
  uppercase={false}
  variant={null}
>
  <div
    aria-checked={true}
    className="toggle"
    onClick={[MockFunction]}
    onKeyPress={[MockFunction]}
    role="checkbox"
    tabIndex={0}
  >
    <label
      className="toggle__switch"
    >
      <input
        className="toggle__input"
        type="checkbox"
      />
      <span
        className="toggle__slider"
      />
    </label>
  </div>
</Toggle>
`;
